### PR TITLE
Writing of trace file

### PR DESCRIPTION
**Pull request by gabrielbosio**

### CODE REVIEW

**Review by fmoletta — COMMENTED** [review]
Is the `cairo_trace_struct` file the fibonacci.cairo's trace file generated by the cairo vm? If so, can we rename it to fibonacci.trace to avoid confusions?

File: pkg/vm/vm_core.go [hunk 1]
```diff
@@ -1,11 +1,68 @@
...
 // Runs Cairo assembly and produces an execution trace.
 type VirtualMachine struct {
-	runContext  RunContext
-	currentStep uint
-	segments    memory.MemorySegmentManager
+	runContext     RunContext
+	currentStep    uint
+	Segments       memory.MemorySegmentManager
+	traceRelocated bool
+	Trace          []TraceEntry
```
[the remark this answers is not in the document]
Comment (gabrielbosio) [reply pkg/vm/vm_core.go]: Done, thanks!

File: pkg/vm/trace.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package vm
+
+import "github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+
+type TraceEntry struct {
+	Pc uint
```
Comment (fmoletta) [inline pkg/vm/trace.go]: Pc should be Relocatable too

File: pkg/vm/vm_core.go [hunk 3]
```diff
@@ -1,15 +1,70 @@
...
+	if len(*relocationTable) < 2 {
+		return errors.New("No relocation found for execution segment")
+	}
+	segment1Base := (*relocationTable)[1]
+
+	for _, entry := range v.Trace {
+		v.RelocatedTrace = append(v.RelocatedTrace, RelocatedTraceEntry{
+			Pc: entry.Pc.RelocateAddress(relocationTable),
+			Ap: entry.Ap.RelocateAddress(relocationTable) + segment1Base,
+			Fp: entry.Fp.RelocateAddress(relocationTable) + segment1Base,
```
Comment (fmoletta) [inline pkg/vm/vm_core.go]: Isn't this adding the segment1Base twice? (given that it is already added in `RelocateAddress`)

Comment (gabrielbosio) [reply pkg/vm/vm_core.go]: I can't find where it's added, because `RelocateAddress` only returns `relocationTable[segmentIndex] + offset`.

Comment (gabrielbosio) [reply pkg/vm/vm_core.go]: The `relocationTable` values have added the first address of the segment 0, which is `1`. In this case, I added to `Ap` and `Fp` the first address of the segment 1.

Comment (fmoletta) [reply pkg/vm/vm_core.go]: `Ap` & `Fp` should always have segment index 1, I think the test values might be the problem here

Comment (fmoletta) [reply pkg/vm/vm_core.go]: Remember when porting tests from cairo-vm, that segment_index is asumed to be 0 for pc and 1 for ap and fp

Comment (gabrielbosio) [reply pkg/vm/vm_core.go]: Ok, so in this case if we take the trace entries from cairo-vm tests, we should add 1 to the ap and fp of the expected entries

Comment (fmoletta) [reply pkg/vm/vm_core.go]: When creating the initial entry (aka the non relocated trace) you should convert each register to Relocatable doing this conversion:
cairo-vm (rust)
TraceEntry {
  pc: pc_offset
  ap: ap_offset
  fp: fp_offset
  }
  
  cairo-vm.go
  TraceEntry {
  pc: 0: pc_offset
  ap : 1: ap_offset
  fp: 1: fp_offset
  }
  Which is adding the correct segment index for each register, that is omitted in the cairo-vm trace for perfromance.
  Values in the RelocatedTrace should be left untouched

**Review by fmoletta — APPROVED** [review]
🚀

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**